libhb/decavcodec.c

```diff
@@ -1611,38 +1611,97 @@ static int decavcodecvInit( hb_work_object_t * w, hb_job_t * job )
 
 static int setup_extradata( hb_work_private_t * pv, AVCodecContext * context )
 {
-    // we can't call the avstream funcs but the read_header func in the
-    // AVInputFormat may set up some state in the AVContext. In particular
-    // vc1t_read_header allocates 'extradata' to deal with header issues
-    // related to Microsoft's bizarre engineering notions. We alloc a chunk
-    // of space to make vc1 work then associate the codec with the context.
-    if (context->extradata == NULL)
+    const AVBitStreamFilter * bsf;
+    AVBSFContext            * ctx = NULL;
+    int                       ii, ret;
+    AVPacket                * avp = pv->pkt;
+    const enum AVCodecID    * ids;
+
+    if (context->extradata != NULL)
+    {
+        return 0;
+    }
+    bsf = av_bsf_get_by_name("extract_extradata");
+    if (bsf == NULL)
+    {
+        hb_error("setup_extradata: bitstream filter lookup failure");
+        return 0;
+    }
+    if (bsf->codec_ids == NULL)
     {
-        if (pv->parser == NULL || pv->parser->parser == NULL ||
-            pv->parser->parser->split == NULL)
+        hb_error("setup_extradata: extract_extradata missing codec_ids");
+        return 0;
+    }
+    for (ids = bsf->codec_ids; *ids != AV_CODEC_ID_NONE; ids++)
+    {
+        if (*ids == context->codec_id)
         {
-            return 0;
+            break;
         }
-        else
+    }
+    if (*ids == AV_CODEC_ID_NONE)
+    {
+        // Codec not supported by extract_extradata BSF
+        return 0;
+    }
+    ret = av_bsf_alloc(bsf, &ctx);
+    if (ret < 0)
+    {
+        hb_error("setup_extradata: bitstream filter alloc failure");
+        return 0;
+    }
+    avcodec_parameters_from_context(ctx->par_in, context);
+    ret = av_bsf_init(ctx);
+    if (ret < 0)
+    {
+        hb_error("setup_extradata: bitstream filter init failure");
+        av_bsf_free(&ctx);
+        return 0;
+    }
+
+    avp->data = pv->packet_info.data;
+    avp->size = pv->packet_info.size;
+    avp->pts  = pv->sequence;
+    avp->dts  = pv->sequence;
+    ret = av_bsf_send_packet(ctx, avp);
```

Comment:
Hint: The BSF API works with refcounted packets and given that this packet here is not refcounted, the data will be copied to a freshly allocated buffer. If you have an AVBufferRef owning the data, you could avoid this. But given that this is only executed for the first few frames, it probably won't matter much.